monitoring/influxdb.go

```diff
@@ -105,52 +210,141 @@ type influxdbMonitor struct {
 	client   influxdb.Client
 	database string
 	logger   log.Logger
+
+	pointChan          chan *influxdb.Point
+	batchWriteInterval time.Duration
+	cacheEvent         *cacheEvent
 }
 
-// InsertRecord part of monitor.Monitor.
-func (im influxdbMonitor) InsertRecord(measurement string, value interface{}, tags map[string]string, fields map[string]interface{}, at time.Time) {
-	if fields == nil {
-		fields = map[string]interface{}{}
+type cacheEvent struct {
+	cacheEvents        int
+	maxCacheEvents     int
+	currentCacheEvents int
+}
+
+func newCacheEvent(cacheEvents int, maxCacheEvents int) *cacheEvent {
+	return &cacheEvent{
+		cacheEvents:        cacheEvents,
+		maxCacheEvents:     maxCacheEvents,
+		currentCacheEvents: cacheEvents,
 	}
+}
 
-	fields["value"] = value
+func (ce *cacheEvent) CurrentCacheEvents() int {
+	return ce.currentCacheEvents
+}
+
+func (ce *cacheEvent) IncreaseCacheEvents() {
+	currEvents := ce.currentCacheEvents + ce.cacheEvents
+	if currEvents > ce.maxCacheEvents {
+		ce.currentCacheEvents = ce.maxCacheEvents
+	} else {
+		ce.currentCacheEvents = currEvents
+	}
+}
+
+func (ce *cacheEvent) Reset() {
+	// This if is rarely true, so add it.
+	if ce.currentCacheEvents != ce.cacheEvents {
+		ce.currentCacheEvents = ce.cacheEvents
+	}
+}
+
+func (im influxdbMonitor) batchWriteTicker() {
+	var points []*influxdb.Point
+	t := time.NewTicker(im.batchWriteInterval)
```

Comment:
`time.Tick` gives you the channel directly. You don't call `time.Ticker.Stop` anyway so you only need the channel.

---

I don't think this is a big problem, but you should consider using `time.After` instead.

`time.Ticker` doesn't work as a "timeout", and using `time.Ticker` will result in this sometimes:

1. read point from channel at time T, reach maximum point count, send points to InfluxDB

2. read point from channel at time T + 10ms, add point to buffer

3. read from ticker channel T + 20ms (or something similar), send a HTTP request with one point

=> "Send data to InfluxDB when we have `MaxCacheEvents` *or* every `batchWriteInterval` seconds

Vs:

1. read point from channel at time T, reach maximum point count, send points to InfluxDB, reset timeout

2. read point from channel at time T + 10ms, add point to buffer

3. can't read from timeout channel until `T + im.batchWriteInterval`

=> "Send data to InfluxDB when we have `MaxCacheEvents` *or* we haven't sent data since `batchWriteInterval` seconds ago"

So if you have a lot of data to send (> ~5000 points/minute), the timeout will never need to trigger.

Reply:
Good point.

Changed `time.Ticker` to `time.After` in https://github.com/theplant/appkit/pull/25/commits/18f119747a2cb76ede9d7f36579a3b7e9fcd5758.